crates/core/tedge_api/src/plugin.rs

```diff
@@ -0,0 +1,619 @@
+    /// # impl tedge_api::Plugin for HeartbeatPlugin {
+    /// #     async fn setup(&mut self) -> Result<(), PluginError> {
+    /// #         unimplemented!()
+    /// #     }
+    /// #
+    /// #     async fn shutdown(&mut self) -> Result<(), PluginError> {
+    /// #         unimplemented!()
+    /// #     }
+    /// # }
+    ///
+    /// println!("{:#?}", HandleTypes::declare_handlers_for::<(Heartbeat,), HeartbeatPlugin>());
+    /// // This will print something akin to:
+    /// //
+    /// // HandleTypes(
+    /// //  [
+    /// //      (
+    /// //          "rust_out::main::_doctest_main_src_plugin_rs_102_0::Heartbeat",
+    /// //          TypeId {
+    /// //              t: 15512189350087767644,
+    /// //          },
+    /// //      ),
+    /// //  ],
+    /// // )
+    /// ```
+    ///
+    /// Should you ask for messages the plugin does _not_ support, you will receive a compile
+    /// error:
+    /// ```compile_fail
+    /// # use async_trait::async_trait;
+    /// # use tedge_api::plugin::{Message, Handle, HandleTypes};
+    /// # use tedge_api::PluginError;
+    ///
+    /// struct Heartbeat;
+    ///
+    /// impl Message for Heartbeat {}
+    ///
+    /// struct HeartbeatPlugin;
+    ///
+    /// // This will fail to compile as the `Heartbeat` message is not handled by the plugin
+    /// println!("{:#?}", HandleTypes::get_handlers_for::<(Heartbeat,), HeartbeatPlugin>());
+    /// ```
+    ///
+    /// The error from the rust compiler would look like this (giving a clear indication of what is
+    /// missing):
+    /// ```text
+    /// error[E0277]: the trait bound `HeartbeatPlugin: Handle<Heartbeat>` is not satisfied
+    ///    --> src/plugin.rs:XX:YYY
+    ///     |
+    /// XX  | println!("{:#?}", HandleTypes::get_handlers_for::<(Heartbeat,), HeartbeatPlugin>());
+    ///     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Handle<Heartbeat>` is not implemented for `HeartbeatPlugin`
+    /// ```
+    pub fn declare_handlers_for<M: MessageBundle, Plugin: DoesHandle<M>>() -> HandleTypes {
+        HandleTypes(M::get_ids())
+    }
+
+    /// Empty list of types. A plugin that does not handle anything will not be able to receive
+    /// messages except through replies sent with [`Reply`](crate::address::Reply)
+    pub fn empty() -> HandleTypes {
+        HandleTypes(Vec::with_capacity(0))
+    }
+}
+
+impl From<HandleTypes> for HashSet<(&'static str, TypeId)> {
+    fn from(ht: HandleTypes) -> Self {
+        ht.0.into_iter().collect()
+    }
+}
+
+/// A thing that can be send around
+///
+/// This trait is a marker trait for all types that can be used as messages which can be send
+/// between plugins in thin-edge.
+pub trait Message: 'static + Send + std::fmt::Debug {
+    type Reply: Message;
+}
+
+/// A bundle of messages
+///
+/// This trait is implemented on types that represent a bundle of different types of messages.
+pub trait MessageBundle {
```

Comment:
Consider renaming this, since a `MessageBundle` is only meant to be used to declare what kind of messages a plugin _handles_. So maybe `HandledBundle` or so?